### PR TITLE
Fixed a serious bug in gmf_ebrisk

### PR DESCRIPTION
**Pull request by micheles**
While working at https://github.com/gem/oq-engine/issues/2017 I discovered a serious bug in the gmf_ebrisk calculator, which is giving completely bogus numbers depending on the number of spawned tasks. This PR fixes the bug and includes some refactoring of the risk calculators.

### CODE REVIEW

File: openquake/calculators/base.py [hunk 1]
```diff
@@ -404,13 +407,7 @@ def basic_pre_execute(self):
         self.read_risk_data()
         if 'source' in oq.inputs:
             wakeup_pool()  # fork before reading the source model
-            if oq.hazard_calculation_id:  # already stored csm
-                logging.info('Reusing composite source model of calc #%d',
-                             oq.hazard_calculation_id)
```
Comment (micheles) [inline openquake/calculators/base.py]: This was dead code

File: openquake/calculators/base.py [hunk 2]
```diff
@@ -306,7 +305,7 @@ def before_export(self):
         """
         Collect the realizations and set the attributes nbytes
         """
-        if hasattr(self, 'rlzs_assoc'):
+        if 'csm_info' in self.datastore and hasattr(self, 'rlzs_assoc'):
```
Comment (ptormene) [inline openquake/calculators/base.py]: what happens otherwise?

File: openquake/calculators/event_based_risk.py [hunk 3]
```diff
@@ -131,14 +133,14 @@ def event_based_risk(riskinput, riskmodel, param, monitor):
...
     R = riskinput.hazard_getter.num_rlzs
     param['lrs_dt'] = numpy.dtype([('rlzi', U16), ('ratios', (F32, (L * I,)))])
     agg = numpy.zeros((E, R, L * I), F32)
     result = dict(assratios=[], lrs_idx=AccumDict(accum=[]), aids=aids)
     if param['avg_losses']:
-        # dict (l, r) -> A losses
-        result['avglosses'] = AccumDict(accum=numpy.zeros(A, F64))
+        # dict (l, r) -> loss_by_aid
+        result['avglosses'] = AccumDict(accum=numpy.zeros(len(assetcol), F64)
+                                        if aids is None else {})
```
Comment (ptormene) [inline openquake/calculators/event_based_risk.py]: Please add a comment describing the different cases

Comment (micheles) [reply openquake/calculators/event_based_risk.py]: Done

File: openquake/calculators/base.py (hunk 2 above)
Comment (micheles) [reply openquake/calculators/base.py]: Nothing bad, I have added a comment

**Pull request merged**